src/services/refactors/moveToFile.ts

```diff
@@ -0,0 +1,1114 @@
+        Debug.assert(actionName === refactorNameForMoveToFile, "Wrong refactor invoked");
+        const statements = Debug.checkDefined(getStatementsToMove(context));
+        Debug.assert(newFile, "Target file does not exist");
+        const edits = textChanges.ChangeTracker.with(context, t => doChange(context, context.file, newFile, context.program, statements, t, context.host, context.preferences));
+        return { edits, renameFilename: undefined, renameLocation: undefined };
+    }
+});
+
+function doChange(context: RefactorContext, oldFile: SourceFile, newFile: string, program: Program, toMove: ToMove, changes: textChanges.ChangeTracker, host: LanguageServiceHost, preferences: UserPreferences): void {
+    const checker = program.getTypeChecker();
+    const usage = getUsageInfo(oldFile, toMove.all, checker);
+    //creating a new file
+    if (!host.fileExists(newFile) || host.fileExists(newFile) && program.getSourceFile(newFile)?.statements.length === 0) {
+        changes.createNewFile(oldFile, newFile, getNewStatementsAndRemoveFromOldFile(oldFile, newFile, usage, changes, toMove, program, host, newFile, preferences, /*newFileExists*/ false));
+        addNewFileToTsconfig(program, changes, oldFile.fileName, newFile, hostGetCanonicalFileName(host));
+    }
+    else {
+        const sourceFile = program.getSourceFile(newFile);
+        if (sourceFile) {
+            const importAdder = codefix.createImportAdder(sourceFile, context.program, context.preferences, context.host);
+            getNewStatementsAndRemoveFromOldFile(oldFile, newFile, usage, changes, toMove, program, host, newFile, preferences, /*newFileExists*/ true, importAdder);
+        }
+    }
+}
+
+function getNewStatementsAndRemoveFromOldFile(
+    oldFile: SourceFile, newFile: string, usage: UsageInfo, changes: textChanges.ChangeTracker, toMove: ToMove, program: Program, host: LanguageServiceHost, newFilename: string, preferences: UserPreferences, newFileExists: boolean, importAdder?: codefix.ImportAdder
+) {
+    const checker = program.getTypeChecker();
+    const prologueDirectives = takeWhile(oldFile.statements, isPrologueDirective);
+    if (oldFile.externalModuleIndicator === undefined && oldFile.commonJsModuleIndicator === undefined && usage.oldImportsNeededByNewFile.size === 0 && usage.newFileImportsFromOldFile.size === 0 && !newFileExists) {
+        deleteMovedStatements(oldFile, toMove.ranges, changes);
+        return [...prologueDirectives, ...toMove.all];
+    }
+
+    let useEsModuleSyntax = true;
+    if (newFileExists) {
+        const newFilePath = program.getSourceFile(newFile)?.path;
+        if (extensionFromPath(newFile) === Extension.Js && newFilePath && getImpliedNodeFormatForFile(newFilePath, /*packageJsonInfoCache*/ undefined, host, program.getCompilerOptions()) !== ModuleKind.ESNext && (oldFile.commonJsModuleIndicator || program.getCompilerOptions().verbatimModuleSyntax)) {

```

Comment:
Bugs and clarity nits:

1. `Js` isn’t the only JS extension—use `isSourceFileJS(newSourceFile)` or `hasJsFileExtension(newFile)` to cover cjs, mjs, jsx, etc.
2. Use `newSourceFile.impliedNodeFormat` instead of `getImpliedNodeFormatForFile` (the latter gets called once during program construction and gets cached on the former)
3. Since this is the case where the new file already exists, it seems like `newSourceFile.commonJsModuleIndicator` and `externalModuleIndicator` are relevant here.
4. Whatever the logic is, it has to be commented since it’s extremely hard to think about.
5. What about the case where the new file doesn’t exist? In that case, we can’t look at the new file’s existing syntax, but the other considerations still apply. There may need to be a test or two for that.

```ts
// Cannot use require if not JS
if (isSourceFileJs(newSourceFile) && (
    // Use require if detected CJS format
    newSourceFile.impliedNodeFormat === ModuleKind.CommonJS ||
    !newSourceFile.impliedNodeFormat && (
        // Target file is already using require,
        // or not using import and old file is using require
        newSourceFile.commonJsModuleIndicator ||
        !newSourceFile.externalModuleIndicator && oldSourceFile.commonJsModuleIndicator) ||
    // Must use require in CJS emit with verbatimModuleSyntax
    getEmitModuleKind(compilerOptions) === ModuleKind.CommonJS && compilerOptions.verbatimModuleSyntax
)) {
  useEsModuleSyntax = false;
}    
```